### DISCUSSION TITLE
How do I add more than one dynamic bond while my simulation runs?

### DISCUSSION THREAD

**Discussion by anushkaj11** [glotzerlab/hoomd-blue#1890] [Support] [answered]
I am trying to add dynamic bonds using the `CustomWriter` class. 

```
class WriteBonds(hoomd.custom.Action):
    def __init__(self, dt):
        self.dt = dt
        self.iter = 0

    def act(self, timestep):
        print(timestep)
        snap = self._state.get_snapshot()
        snap.bonds.N = snap.bonds.N +1
        snap.bonds.group[-1] = [self.iter+1, self.iter+2]
        print("snap.bonds.group: ", snap.bonds.group)
        self._state.set_snapshot(snap) 
        self.iter = self.iter +1
```

 I created a `WriteBonds` object and call it during the simulation with 

```
    write_bonds = WriteBonds(dt)                             
    bond_writer_op = hoomd.write.CustomWriter(action=write_bonds, trigger=hoomd.trigger.Periodic(10000))
    sim.operations.writers.append(bond_writer_op) 
```

My issue is, I am only able to add bonds at the first trigger. I checked if my `snap.bonds.group`  was getting updated and it seems to be getting bigger with every trigger event but I don't see more added bonds when the simulation ends. 

Am I missing something here? I also tried using `CustomUpdater` but that doesn't seem to work to add more bonds.
I also noticed that my modifications are being written at the start - not dynamically.

UPDATE: it seems I needed to add `'topology'` to `dynamic` in `hoomd.write.GSD` - I now see the kind of dynamic bonding I was looking for (Big thanks to my colleague who spotted this!)
The code above seems to be working fine.

FOLLOW-UP: I still don't know the difference between `CustomWriter` and `CustomUpdater` as it pertains to adding bonds in my code, both seem to be working just fine for adding bonds and applying forces on the new bonds

**Comment by joaander** [comment, marked as answer]
`Writers` and `Updaters` execute at different points during the `Simulation.run` loop: See https://hoomd-blue.readthedocs.io/en/v4.8.2/package-hoomd.html#hoomd.Simulation.run for details.

You should use an `Updater` when you *modify* system state and `Writer` when you *read* the system state and write results to an output (e.g. file or stream).
